File: monitoring-backend/app/schemas/location.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class LocationGroupBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class LocationGroupCreate(LocationGroupBase):
    pass


class LocationGroupUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None


class LocationGroupResponse(LocationGroupBase):
    group_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LocationCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255, description="Location name")
    address: str = Field(..., description="Location address")
    location_type: str = Field(..., description="Location type")
    latitude: float = Field(..., description="Location latitude")
    longitude: float = Field(..., description="Location longitude")
    description: Optional[str] = None
    group_id: Optional[int] = None


class LocationUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    address: Optional[str] = None
    location_type: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    description: Optional[str] = None
    group_id: Optional[int] = None


class LocationResponse(BaseModel):
    location_id: int
    name: str
    address: str
    location_type: str
    latitude: float
    longitude: float
    description: Optional[str]
    group_id: Optional[int]
    group_name: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LocationOptionResponse(BaseModel):
    location_id: int
    name: str
    location_type: str
    location_type_label: str
    group_id: Optional[int]
    group_name: Optional[str]


class LocationPageResponse(BaseModel):
    items: list[LocationResponse]
    total: int
    page: int
    page_size: int

File: monitoring-frontend/lib/models/location.dart
class Location {
  final int id;
  final double latitude;
  final double longitude;
  final String? address;
  final String? type;
  final String name;
  final String? description;
  final int? groupId;
  final String? groupName;
  final String? typeLabel;

  Location(
      {required this.id,
      required this.latitude,
      required this.longitude,
      this.address,
      this.type,
      required this.name,
      this.description,
      this.groupId,
      this.groupName,
      this.typeLabel});

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      id: json['location_id'],
      name: json['name'],
      latitude: json['latitude'],
      longitude: json['longitude'],
      address: json['address'],
      type: json['location_type'],
      description: json['description'],
      groupId: json['group_id'],
      groupName: json['group_name'],
      typeLabel: json['location_type_label'],
    );
  }
  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'latitude': latitude,
      'longitude': longitude,
      'address': address,
      'location_type': type,
      'description': description,
    };
  }
}

class LocationGroup {
  final int groupId;
  final String name;
  final String? description;

  LocationGroup({
    required this.groupId,
    required this.name,
    this.description,
  });

  factory LocationGroup.fromJson(Map<String, dynamic> json) {
    return LocationGroup(
      groupId: json['group_id'],
      name: json['name'],
      description: json['description'],
    );
  }
}

class LocationPage {
  final List<Location> items;
  final int total;
  final int page;
  final int pageSize;

  LocationPage({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory LocationPage.fromJson(Map<String, dynamic> json) {
    final raw = json['items'] as List? ?? [];
    return LocationPage(
      items: raw.map((e) => Location.fromJson(e)).toList(),
      total: json['total'] ?? 0,
      page: json['page'] ?? 1,
      pageSize: json['page_size'] ?? 10,
    );
  }
}
